File: go/builders_test.go
package afi

import (
	"bytes"
	"math/big"
	"testing"

	"github.com/ethereum/go-ethereum/common"
)

func TestBuildUniV3Step_ByteLength(t *testing.T) {
	step, err := BuildUniV3Step(
		common.HexToAddress("0x1111111111111111111111111111111111111111"),
		3000,
		big.NewInt(1_000_000),
		big.NewInt(0),
	)
	if err != nil {
		t.Fatalf("unexpected: %v", err)
	}
	if len(step.Data) != 59 {
		t.Errorf("UniV3 expected 59 bytes, got %d", len(step.Data))
	}
	if step.ID != RouteIDUniV3 {
		t.Errorf("UniV3 id: got %d want %d", step.ID, RouteIDUniV3)
	}
}

func TestBuildCakeV3Step_ByteLength(t *testing.T) {
	step, err := BuildCakeV3Step(
		common.HexToAddress("0x1111111111111111111111111111111111111111"),
		500,
		big.NewInt(1),
		big.NewInt(0),
	)
	if err != nil {
		t.Fatalf("unexpected: %v", err)
	}
	if len(step.Data) != 59 {
		t.Errorf("CakeV3 expected 59 bytes, got %d", len(step.Data))
	}
	if step.ID != RouteIDCakeV3 {
		t.Errorf("CakeV3 id: got %d want %d", step.ID, RouteIDCakeV3)
	}
}

func TestBuildUniV4Step_ByteLength(t *testing.T) {
	step, err := BuildUniV4Step(
		common.HexToAddress("0x1111111111111111111111111111111111111111"),
		common.HexToAddress("0x2222222222222222222222222222222222222222"),
		3000,
		60,
		common.Address{},
		true,
		big.NewInt(1),
	)
	if err != nil {
		t.Fatalf("unexpected: %v", err)
	}
	if len(step.Data) != 83 {
		t.Errorf("UniV4 expected 83 bytes, got %d", len(step.Data))
	}
	if step.ID != RouteIDUniV4 {
		t.Errorf("UniV4 id: got %d want %d", step.ID, RouteIDUniV4)
	}
	// zeroForOne flag is at offset 66.
	if step.Data[66] != 1 {
		t.Errorf("UniV4 zeroForOne byte: got %d want 1", step.Data[66])
	}
}

func TestBuildAerodromeStep_ByteLength(t *testing.T) {
	step, err := BuildAerodromeStep(
		common.HexToAddress("0x1111111111111111111111111111111111111111"),
		200,
		big.NewInt(0),
		big.NewInt(1),
	)
	if err != nil {
		t.Fatalf("unexpected: %v", err)
	}
	if len(step.Data) != 59 {
		t.Errorf("Aerodrome expected 59 bytes, got %d", len(step.Data))
	}
	if step.ID != RouteIDAerodrome {
		t.Errorf("Aerodrome id: got %d want %d", step.ID, RouteIDAerodrome)
	}
}

func TestBuildAerodromeStep_NegativeTickSpacing(t *testing.T) {
	step, err := BuildAerodromeStep(
		common.HexToAddress("0x1111111111111111111111111111111111111111"),
		-1, // sign-extend to 0xFFFFFF
		big.NewInt(0),
		big.NewInt(1),
	)
	if err != nil {
		t.Fatalf("unexpected: %v", err)
	}
	// tickSpacing lives at offset 20..22.
	wantTick := []byte{0xFF, 0xFF, 0xFF}
	if !bytes.Equal(step.Data[20:23], wantTick) {
		t.Errorf("tickSpacing two's-complement: got %x want %x", step.Data[20:23], wantTick)
	}

	step2, err := BuildAerodromeStep(
		common.HexToAddress("0x1111111111111111111111111111111111111111"),
		-2, // -> 0xFFFFFE
		big.NewInt(0),
		big.NewInt(1),
	)
	if err != nil {
		t.Fatalf("unexpected: %v", err)
	}
	wantTick2 := []byte{0xFF, 0xFF, 0xFE}
	if !bytes.Equal(step2.Data[20:23], wantTick2) {
		t.Errorf("tickSpacing -2 two's-complement: got %x want %x", step2.Data[20:23], wantTick2)
	}
}

func TestBuildBalancerV3Step_ByteLength(t *testing.T) {
	step, err := BuildBalancerV3Step(
		common.HexToAddress("0x1111111111111111111111111111111111111111"),
		common.HexToAddress("0x2222222222222222222222222222222222222222"),
		big.NewInt(1),
	)
	if err != nil {
		t.Fatalf("unexpected: %v", err)
	}
	if len(step.Data) != 56 {
		t.Errorf("BalancerV3 expected 56 bytes, got %d", len(step.Data))
	}
	if step.ID != RouteIDBalancerV3 {
		t.Errorf("BalancerV3 id: got %d want %d", step.ID, RouteIDBalancerV3)
	}
}

func TestBuildFluidStep_ByteLength(t *testing.T) {
	step, err := BuildFluidStep(
		common.HexToAddress("0x1111111111111111111111111111111111111111"),
		true,
		common.HexToAddress("0x2222222222222222222222222222222222222222"),
		big.NewInt(1),
	)
	if err != nil {
		t.Fatalf("unexpected: %v", err)
	}
	if len(step.Data) != 57 {
		t.Errorf("Fluid expected 57 bytes, got %d", len(step.Data))
	}
	if step.ID != RouteIDFluid {
		t.Errorf("Fluid id: got %d want %d", step.ID, RouteIDFluid)
	}
	if step.Data[20] != 1 {
		t.Errorf("Fluid swap0to1 byte: got %d want 1", step.Data[20])
	}
}

func TestBuildCurve128Step_ByteLength(t *testing.T) {
	step, err := BuildCurve128Step(
		0, 1,
		big.NewInt(1),
		common.HexToAddress("0x1111111111111111111111111111111111111111"),
		common.HexToAddress("0x2222222222222222222222222222222222222222"),
	)
	if err != nil {
		t.Fatalf("unexpected: %v", err)
	}
	if len(step.Data) != 58 {
		t.Errorf("Curve128 expected 58 bytes, got %d", len(step.Data))
	}
	if step.ID != RouteIDCurve128 {
		t.Errorf("Curve128 id: got %d want %d", step.ID, RouteIDCurve128)
	}
}

func TestBuildCurve256Step_ByteLength(t *testing.T) {
	step, err := BuildCurve256Step(
		1, 0,
		big.NewInt(1),
		common.HexToAddress("0x1111111111111111111111111111111111111111"),
		common.HexToAddress("0x2222222222222222222222222222222222222222"),
	)
	if err != nil {
		t.Fatalf("unexpected: %v", err)
	}
	if len(step.Data) != 58 {
		t.Errorf("Curve256 expected 58 bytes, got %d", len(step.Data))
	}
	if step.ID != RouteIDCurve256 {
		t.Errorf("Curve256 id: got %d want %d", step.ID, RouteIDCurve256)
	}
}

func TestBuildAaveLiquidatorStep_ByteLength(t *testing.T) {
	step, err := BuildAaveLiquidatorStep(
		common.HexToAddress("0x1111111111111111111111111111111111111111"),
		common.HexToAddress("0x2222222222222222222222222222222222222222"),
		common.HexToAddress("0x3333333333333333333333333333333333333333"),
	)
	if err != nil {
		t.Fatalf("unexpected: %v", err)
	}
	if len(step.Data) != 60 {
		t.Errorf("AaveLiquidator expected 60 bytes, got %d", len(step.Data))
	}
	if step.ID != RouteIDAaveLiquidator {
		t.Errorf("AaveLiquidator id: got %d want %d", step.ID, RouteIDAaveLiquidator)
	}
}

func TestBuildUniV3Step_MinOutTooBig(t *testing.T) {
	// 2^128 = 1 << 128
	tooBig := new(big.Int).Lsh(big.NewInt(1), 128)
	_, err := BuildUniV3Step(
		common.HexToAddress("0x1111111111111111111111111111111111111111"),
		3000,
		tooBig,
		nil,
	)
	if err == nil {
		t.Error("expected error for minOut > uint128, got nil")
	}
}

func TestBuildBalancerV3Step_NilMinOut(t *testing.T) {
	_, err := BuildBalancerV3Step(
		common.HexToAddress("0x1111111111111111111111111111111111111111"),
		common.HexToAddress("0x2222222222222222222222222222222222222222"),
		nil,
	)
	if err == nil {
		t.Error("expected error for nil minOut, got nil")
	}
}

File: nodejs/src/__tests__/builders.test.ts
import { describe, expect, it } from "vitest"
import { size } from "viem"
import {
  ROUTE_ID,
  buildAaveLiquidatorStep,
  buildAerodromeStep,
  buildBalancerV3Step,
  buildCakeV3Step,
  buildCurve128Step,
  buildCurve256Step,
  buildFluidStep,
  buildUniV3Step,
  buildUniV4Step,
  stepDataByteLength,
} from "../builders.js"

const TOKEN_OUT = "0x4200000000000000000000000000000000000006" as const
const POOL      = "0x1111111111111111111111111111111111111111" as const
const USER      = "0x2222222222222222222222222222222222222222" as const
const COLLATERAL = "0x3333333333333333333333333333333333333333" as const
const HOOKS     = "0x0000000000000000000000000000000000000000" as const

describe("builders — UniV3", () => {
  it("encodes a 59-byte stepData with id=3", () => {
    const step = buildUniV3Step({ tokenOut: TOKEN_OUT, fee: 500, minOut: 1n })
    expect(step.id).toBe(ROUTE_ID.UNI_V3)
    expect(step.id).toBe(3)
    expect(size(step.data)).toBe(59)
    // 20 (addr) + 3 (fee=0x0001f4) + 16 (minOut=0x..01) + 20 (sqrt=0)
    expect(step.data.slice(2, 2 + 40)).toBe(TOKEN_OUT.slice(2).toLowerCase())
    expect(step.data.slice(2 + 40, 2 + 40 + 6)).toBe("0001f4")
  })

  it("uses 0 as default sqrtPriceLimit", () => {
    const step = buildUniV3Step({ tokenOut: TOKEN_OUT, fee: 0, minOut: 0n })
    // Last 20 bytes (40 hex chars) should all be zero.
    expect(step.data.slice(-40)).toBe("0".repeat(40))
  })
})

describe("builders — CakeV3", () => {
  it("encodes a 59-byte stepData with id=2", () => {
    const step = buildCakeV3Step({ tokenOut: TOKEN_OUT, fee: 100, minOut: 1n })
    expect(step.id).toBe(ROUTE_ID.CAKE_V3)
    expect(step.id).toBe(2)
    expect(size(step.data)).toBe(59)
  })
})

describe("builders — UniV4", () => {
  it("encodes an 83-byte stepData with id=4", () => {
    const step = buildUniV4Step({
      currency0:   "0x0000000000000000000000000000000000000000",
      currency1:   TOKEN_OUT,
      fee:         3000,
      tickSpacing: 60,
      hooks:       HOOKS,
      zeroForOne:  true,
      minOut:      1_000n,
    })
    expect(step.id).toBe(ROUTE_ID.UNI_V4)
    expect(step.id).toBe(4)
    expect(size(step.data)).toBe(83)
  })

  it("preserves zeroForOne=false as 0x00 byte", () => {
    const step = buildUniV4Step({
      currency0:   TOKEN_OUT,
      currency1:   "0x0000000000000000000000000000000000000000",
      fee:         3000,
      tickSpacing: 60,
      hooks:       HOOKS,
      zeroForOne:  false,
      minOut:      1n,
    })
    // zeroForOne byte sits at offset 66 (= 20+20+3+3+20)
    expect(step.data.slice(2 + 66 * 2, 2 + 66 * 2 + 2)).toBe("00")
  })
})

describe("builders — Aerodrome", () => {
  it("encodes a 59-byte stepData with id=5", () => {
    const step = buildAerodromeStep({ tokenOut: TOKEN_OUT, tickSpacing: 50, minOut: 1n })
    expect(step.id).toBe(ROUTE_ID.AERODROME)
    expect(step.id).toBe(5)
    expect(size(step.data)).toBe(59)
  })

  it("encodes negative tickSpacing as signed int24 (two's complement)", () => {
    // -1 in int24 = 0xffffff
    const step = buildAerodromeStep({ tokenOut: TOKEN_OUT, tickSpacing: -1, minOut: 0n })
    // tickSpacing byte slot: offset 20..23
    const slot = step.data.slice(2 + 20 * 2, 2 + 23 * 2)
    expect(slot).toBe("ffffff")
  })
})

describe("builders — BalancerV3", () => {
  it("encodes a 56-byte stepData with id=9", () => {
    const step = buildBalancerV3Step({ pool: POOL, tokenOut: TOKEN_OUT, minOut: 42n })
    expect(step.id).toBe(ROUTE_ID.BALANCER_V3)
    expect(step.id).toBe(9)
    expect(size(step.data)).toBe(56)
  })
})

describe("builders — Fluid", () => {
  it("encodes a 57-byte stepData with id=7", () => {
    const step = buildFluidStep({ pool: POOL, swap0to1: true, tokenOut: TOKEN_OUT, minOut: 1n })
    expect(step.id).toBe(ROUTE_ID.FLUID)
    expect(step.id).toBe(7)
    expect(size(step.data)).toBe(57)
    // swap0to1=true at offset 20
    expect(step.data.slice(2 + 20 * 2, 2 + 21 * 2)).toBe("01")
  })

  it("encodes swap0to1=false as 0x00 byte", () => {
    const step = buildFluidStep({ pool: POOL, swap0to1: false, tokenOut: TOKEN_OUT, minOut: 1n })
    expect(step.data.slice(2 + 20 * 2, 2 + 21 * 2)).toBe("00")
  })
})

describe("builders — Curve128", () => {
  it("encodes a 58-byte stepData with id=1", () => {
    const step = buildCurve128Step({ i: 0, j: 1, minDy: 0n, pool: POOL, tokenOut: TOKEN_OUT })
    expect(step.id).toBe(ROUTE_ID.CURVE128)
    expect(step.id).toBe(1)
    expect(size(step.data)).toBe(58)
    expect(step.data.slice(2, 2 + 4)).toBe("0001")
  })
})

describe("builders — Curve256", () => {
  it("encodes a 58-byte stepData with id=6", () => {
    const step = buildCurve256Step({ i: 0, j: 2, minDy: 0n, pool: POOL, tokenOut: TOKEN_OUT })
    expect(step.id).toBe(ROUTE_ID.CURVE256)
    expect(step.id).toBe(6)
    expect(size(step.data)).toBe(58)
  })
})

describe("builders — AaveLiquidator", () => {
  it("encodes a 60-byte stepData with id=8", () => {
    const step = buildAaveLiquidatorStep({ pool: POOL, user: USER, collateralAsset: COLLATERAL })
    expect(step.id).toBe(ROUTE_ID.AAVE_LIQUIDATOR)
    expect(step.id).toBe(8)
    expect(size(step.data)).toBe(60)
  })
})

describe("stepDataByteLength", () => {
  it("returns the byte length of the data payload", () => {
    const step = buildUniV3Step({ tokenOut: TOKEN_OUT, fee: 500, minOut: 1n })
    expect(stepDataByteLength(step.data)).toBe(59)
  })
})
